python/ray/dag/compiled_dag_node.py

```diff
@@ -858,6 +864,76 @@ def _get_node_id(self):
         self._dag_submitter.start()
         self._dag_output_fetcher.start()
 
+    def _verify_graph(self) -> bool:
+        """
+        Create a graph following these rules:
+
+        1. Add an edge from task_{bind_index} to task_{bind_index+1}
+           on the same actor.
+        2. Add an edge from the writer to the reader if the channel
+           isn't an NCCL channel.
+        3. Add an edge from the writer and reader of an NCCL channel
+           to the node that has the next bind index on the same actor
+           as the writer.
+
+        Use topological sort to verify whether the graph is a DAG.
+        If not, the DAG will result in a deadlock.
+        """
+
+        class GraphNode:
+            def __init__(self):
+                self.in_edges = set()
+                self.out_edges = set()
+
+            @property
+            def in_degree(self) -> int:
+                return len(self.in_edges)
+
+        from ray.dag import ClassMethodNode
+
+        def _get_next_task_idx(task: "CompiledTask") -> Optional[int]:
+            if not isinstance(task.dag_node, ClassMethodNode):
+                return None
+            actor_handle = task.dag_node._get_actor_handle()
+            bind_index = task.dag_node._get_bind_index()
+            for same_node_task in self.actor_to_tasks[actor_handle]:
+                if same_node_task.dag_node._get_bind_index() == bind_index + 1:
+                    return same_node_task.idx
+            return None
+
+        def _add_edge(
+            graph: Dict[int, GraphNode], from_idx: int, to_idx: Optional[int]
+        ):
+            if to_idx is None:
+                return
+            graph[from_idx].out_edges.add(to_idx)
+            graph[to_idx].in_edges.add(from_idx)
+
+        graph = defaultdict(GraphNode)
+        for idx, task in self.idx_to_task.items():
+            next_task_idx = _get_next_task_idx(task)
+            _add_edge(graph, idx, next_task_idx)
```

Comment:
Please add a comment for each `_add_edge` call, so that we know which case it is.

Reply:
Update https://github.com/ray-project/ray/pull/45960/commits/8267889e8e318dfe57e3119e1a9d769d91f474ee